internal/sourcemap/metadata_fetcher.go

```diff
@@ -0,0 +1,320 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sourcemap
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+const (
+	syncTimeout = 10 * time.Second
+)
+
+type MetadataESFetcher struct {
+	esClient         *elasticsearch.Client
+	index            string
+	set              map[Identifier]string
+	alias            map[Identifier]*Identifier
+	mu               sync.RWMutex
+	logger           *logp.Logger
+	init             chan struct{}
+	invalidationChan chan<- []Identifier
+}
+
+func NewMetadataFetcher(ctx context.Context, esClient *elasticsearch.Client, index string) (MetadataFetcher, <-chan []Identifier) {
+	invalidationCh := make(chan []Identifier)
+
+	s := &MetadataESFetcher{
+		esClient:         esClient,
+		index:            index,
+		set:              make(map[Identifier]string),
+		alias:            make(map[Identifier]*Identifier),
+		logger:           logp.NewLogger(logs.Sourcemap),
+		init:             make(chan struct{}),
+		invalidationChan: invalidationCh,
+	}
+
+	s.startBackgrounSync(ctx)
+
+	return s, invalidationCh
+}
+
+func (s *MetadataESFetcher) GetID(key Identifier) (*Identifier, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if _, ok := s.set[key]; ok {
+		return &key, ok
+	}
+
+	// path is missing from the metadata cache (and ES).
+	// Is it an alias ?
+	// Try to retrieve the sourcemap from the alias map
+	i, ok := s.alias[key]
+	return i, ok
+}
+
+func (s *MetadataESFetcher) Ready() <-chan struct{} {
+	return s.init
+}
+
+func (s *MetadataESFetcher) startBackgrounSync(parent context.Context) {
+	go func() {
+		s.logger.Debug("populating metadata cache")
+
+		// First run, populate cache
+		ctx, cancel := context.WithTimeout(parent, syncTimeout)
+		defer cancel()
+
+		if err := s.sync(ctx); err != nil {
+			s.logger.Errorf("failed to fetch sourcemaps metadata: %v", err)
+		} else {
+			// only close the init chan and mark the fetcher as ready if
+			// sync succeeded
+			close(s.init)
+		}
+
+		s.logger.Info("init routine completed")
```

Comment:
This should either be moved into the `else` close, or rephrased. Right now it leads to
```
{"log.level":"error","@timestamp":"2023-02-02T08:34:43.220+0100","log.logger":"sourcemap","log.origin":{"file.name":"sourcemap/metadata_fetcher.go","file.line":97},"message":"failed to fetch sourcemaps metadata: failed to run initial search query: context deadline exceeded","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-02-02T08:34:43.220+0100","log.logger":"sourcemap","log.origin":{"file.name":"sourcemap/metadata_fetcher.go","file.line":104},"message":"init routine completed","service.name":"apm-server","ecs.version":"1.6.0"}
```

Reply:
Updatd :+1: